tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,863 @@
+	tuplestorecxt->sort_collations = palloc(sizeof(Oid) * n_keys);
+	tuplestorecxt->nulls_first = palloc(sizeof(bool) * n_keys);
+
+	for (int n = 0; n < n_keys; n++)
+		compress_chunk_populate_sort_info_for_column(RelationGetRelid(uncompressed_chunk_rel),
+													 keys[n],
+													 &tuplestorecxt->sort_keys[n],
+													 &tuplestorecxt->sort_operators[n],
+													 &tuplestorecxt->sort_collations[n],
+													 &tuplestorecxt->nulls_first[n]);
+
+	tuplestorecxt->tuplestore = tuplesort_begin_heap(RelationGetDescr(uncompressed_chunk_rel),
+													 n_keys,
+													 tuplestorecxt->sort_keys,
+													 tuplestorecxt->sort_operators,
+													 tuplestorecxt->sort_collations,
+													 tuplestorecxt->nulls_first,
+													 maintenance_work_mem,
+													 NULL,
+													 false);
+	return tuplestorecxt;
+}
+
+/* cleaup tuplesort state */
+static void
+cleanup_tuplestorecxt(CompressTuplesortstateCxt *tuplestorecxt)
+{
+	if (tuplestorecxt)
+	{
+		tuplesort_end(tuplestorecxt->tuplestore);
+		pfree(tuplestorecxt->sort_keys);
+		pfree(tuplestorecxt->sort_operators);
+		pfree(tuplestorecxt->sort_collations);
+		pfree(tuplestorecxt->nulls_first);
+		pfree(tuplestorecxt);
+	}
+}
+
+/*
+ * Helper method to fetch all rows from uncompressed chunk
+ * and sort these rows and save it in tuplestore
+ */
+static int
+save_uncompressed_rows_in_tuplestore(Chunk *uncompressed_chunk,
+									 Tuplesortstate *uncompressed_rows_sortstate)
+{
+	Relation uncompressed_chunk_rel;
+	TableScanDesc heapScan;
+	HeapTuple uncompressed_tuple;
+	TupleTableSlot *heap_tuple_slot = NULL;
+	int total_uncompressed_rows = 0;
+
+	uncompressed_chunk_rel = RelationIdGetRelation(uncompressed_chunk->table_id);
+	heapScan = table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), 0, NULL);
+	heap_tuple_slot =
+		MakeTupleTableSlot(RelationGetDescr(uncompressed_chunk_rel), &TTSOpsHeapTuple);
+
+	while ((uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection)) != NULL)
+	{
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(uncompressed_rows_sortstate, heap_tuple_slot);
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+		total_uncompressed_rows++;
+	}
+	RelationClose(uncompressed_chunk_rel);
+	ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	table_endscan(heapScan);
+	/* sort the tuplestore */
+	tuplesort_performsort(uncompressed_rows_sortstate);
+	return total_uncompressed_rows;
+}
+
+/*
+ * Helper method to fetch _ts_meta_count from a compressed tuple.
+ */
+static void
+fetch_segment_metadata_count(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+							 Datum *metacount_val)
+{
+	Datum *values = palloc0(sizeof(Datum) * decompressor->in_desc->natts);
+	bool *nulls = palloc0(sizeof(bool) * decompressor->in_desc->natts);
+	heap_deform_tuple(compressed_tuple, decompressor->in_desc, values, nulls);
```

Comment:
`slot_getattr` will do that for you, and in a more efficient way (not all columns but only up to the attribute that is required, this is faster in Postgres).